### ISSUE TITLE
[Qt bug] Database write access to windows share with junction links

### ISSUE THREAD

**Issue by Hamzelot** [keepassxreboot/keepassxc#4253]
Hello,

an attempt is made to access the .kdbx database via a Windows network share. This share is provided by a Windows server and is deployed as a network drive on the client. A special feature of the share is that the directory points to another directory within the server via Windows Junctions (https://docs.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions). Opening the database is no problem, only changes to the database cannot be saved. The error message in German is "Schreiben der Datenbank fehlgeschlagen: Die Syntax für den Dateinamen, Verzeichnisnamen oder die Datenträgerbezeichnung ist falsch.".
If the database is opened on a conventional Windows share without junctions, there are no problems when saving.

KeePassXC - Version 2.5.3
Revision: f8c962b

Qt 5.13.2
Diagnostic mode is deactivated.

Operating system: Windows 10 (10.0)
CPU architecture: x86_64
Kernel: winns 10.0.16299

Activated extensions:
- Auto-Type
- Browser integration
- SSH Agent
- KeeShare (confirmed and unconfirmed sharing)
- YubiKey

Cryptographic libraries:
 libgcrypt 1.8.5

**Comment by droidmonkey** [comment]
And all other applications can save to this junction folder without trouble?

**Comment by Hamzelot** [comment]
Hi,

yes I have write access to the directory, but not full access I have deactivated the mentioned option, but it doesn't change anything. In the share I can create and modify Word files, PDF files etc. 

I just re-created the whole thing.

Windows Server 2016 releases a share. In this share there is a link (junction) to a folder on the desktop on the Windows Server. I have created this link with this: 
"mklink /J C:\Junction-Test\junction C:\Users\blabla\Documents\Test-Freigabe\"
The C:\Junction-Test folder is shared and the C:\Test\junction\ folder contains the KeePassXC database, Passwörter.kdbx so the database is actually on Documents\Test-Freigabe.

The user who has access to the folder from Network has full access. 

Now, it's a different error message: "Schreiben der Datenbank fehlgeschlagen: Die Zieldatei kann nicht geöffnet werden: Das System kann den angegebenen Pfad nicht finden."

I replaced server and users details with blabla/blabla-server.

I have checked with the proc-manager what the KeePassXC.exe process does and I notice the following : 

```
8928753 CloseFile	\\blabla-server\Junction-Test\junction	SUCCESS	
9006180 CreateFile C:\Windows\CSC\v2.0.6\namespace\blabla-server NAME NOT FOUND Desired Access: Read EA, Write EA, Read Attributes, Write Attributes, Delete, Read Control, Synchronize, Disposition: Open, Options: Synchronous IO Non-Alert, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9156612 CreateFile \\blabla-server\Junction-Test\\junction SUCCESS Desired Access: Read EA, Read Attributes, Synchronize, Disposition: Open, Options: Synchronous IO Non-Alert, Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
9158701 CreateFile C:\Windows\CSC\v2.0.6\namespace\blabla-server NAME NOT FOUND Desired Access: Read EA, Write EA, Read Attributes, Write Attributes, Delete, Read Control, Synchronize, Disposition: Open, Options: Synchronous IO Non-Alert, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9242708 CreateFile C:\Windows\CSC\v2.0.6\namespace\blabla-server NAME NOT FOUND Desired Access: Read EA, Write EA, Read Attributes, Write Attributes, Delete, Read Control, Synchronize, Disposition: Open, Options: Synchronous IO Non-Alert, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9327075 FileSystemControl \\blabla-server\Junction-Test\junction SUCCESS Control: FSCTL_GET_REPARSE_POINT
9369372 CloseFile \\blabla-server\Junction-Test\junction SUCCESS 
9418436 CreateFile C:\Users SUCCESS Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
9419016 QueryNetworkOpenInformationFile C:\Users SUCCESS CreationTime: 29.09.2017 09:45:11, LastAccessTime: 10.01.2020 07:30:24, LastWriteTime: 10.01.2020 07:30:24, ChangeTime: 10.01.2020 07:30:24, AllocationSize: 01.01.1601 01:00:00, EndOfFile: 01.01.1601 01:00:00, FileAttributes: RD
9419326 CloseFile C:\Users SUCCESS 
9422731 CreateFile C:\Users\blabla NAME NOT FOUND Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9424687 CreateFile C:\Users\blabla\Documents PATH NOT FOUND Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9426631 CreateFile C:\Users\blabla\Documents\Test-Freigabe PATH NOT FOUND Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
9428566 CreateFile C:\Users\blabla\Documents\Test-Freigabe\Passwörter.kdbx PATH NOT FOUND Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
```
The process tries to open the file C:\Users\blabla\Documents\Test-Freigabe\Passwörter.kdbx on my local PC, but it is on the Windows Server.

The same directly in the folder C:\Junction-Test without junction and the same authorization, it work :
```

2159406 QueryDeviceInformationVolume	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	DeviceType: Disk, Characteristics: Remote
2160085 DeviceIoControl	\\blabla-server\Junction-Test\\Passwörter.kdbx	BUFFER OVERFLOW	Control: 0x140410 (Device:0x14 Function:260 Method: 0)
2161299 DeviceIoControl	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	Control: 0x140410 (Device:0x14 Function:260 Method: 0)
2162449 DeviceIoControl	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	Control: 0x140414 (Device:0x14 Function:261 Method: 0)
2205231 QueryDeviceInformationVolume	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	DeviceType: Disk, Characteristics: Remote
2205932 QueryRemoteProtocolInformation	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	
2206732 RegOpenKey	HKLM\Software\Policies\Microsoft\Windows\System	SUCCESS	Desired Access: Query Value
2207166 RegQueryValue	HKLM\SOFTWARE\Policies\Microsoft\Windows\System\CopyFileChunkSize	NAME NOT FOUND	Length: 20
2207511 RegQueryValue	HKLM\SOFTWARE\Policies\Microsoft\Windows\System\CopyFileOverlappedCount	NAME NOT FOUND	Length: 20
2207827 RegCloseKey	HKLM\SOFTWARE\Policies\Microsoft\Windows\System	SUCCESS	
CreationTime: 01.01.1601 01:00:00, LastAccessTime: 01.01.1601 01:00:00, LastWriteTime: 28.01.2020 16:44:06, ChangeTime: 29.01.2020 09:14:59, FileAttributes: n/a
2289903 CloseFile	\\blabla-server\Junction-Test\\Passwörter.kdbx	SUCCESS	

```
Can someone recreate that?

**Comment by droidmonkey** [comment]
This is probably because we request the Canonical File Path from Qt. This is the "resolved path" after all links are followed. Unfortunately it looks like Windows returns the junction folder in this instance even though you are accessing from the network share. This *could* be a Qt bug, but I bet you its a Windows bug/feature. Unfortunately there is nothing we can do to fix this without breaking compatibility with links on Linux and macOS.

**Comment by droidmonkey** [comment]
Confirmed, looks like there were a lot of issues here. This should be corrected once we move to Qt 5.14 on Windows.

https://bugreports.qt.io/browse/QTBUG-73688

**Comment by kjhoerr** [comment]
For reference, I have this same issue when trying to access WSL database files using a UNC filepath (configured as such since FUSE works on WSL 2). In my case it won't even open the file to begin with. It's almost definitely linked to that Qt bug.
